include/util/multi_level_partition.hpp

```diff
@@ -0,0 +1,250 @@
+#ifndef OSRM_UTIL_MULTI_LEVEL_PARTITION_HPP
+#define OSRM_UTIL_MULTI_LEVEL_PARTITION_HPP
+
+#include "util/typedefs.hpp"
+#include "util/for_each_pair.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <cstdint>
+#include <numeric>
+#include <vector>
+
+#include <boost/range/adaptor/reversed.hpp>
+
+namespace osrm
+{
+namespace util
+{
+namespace detail
+{
+// get the msb of an integer
+// return 0 for integers without msb
+template <typename T> std::size_t highestMSB(T value)
```

Comment:
Replaced this with a single instruction, which improves 30x on the loop implementation. Yay!